### PR TITLE
expose snarky hashToGroup

### PR DESCRIPTION
**Pull request by Trivo25**
- exposes Snark_params.Group_map.Checked.to_group
- and implements an out-of-snark variant that is required for mina-signer. [see reference implementation](https://github.com/o1-labs/snarky/blob/78e0d952518f75b5382f6d735adb24eef7a0fa90/group_map/group_map.ml)

note: the params needed for the curve operation are currently hard-coded (originally taken from OCaml) and not calculated in JS

TODO: push bindings 

closes https://github.com/o1-labs/snarkyjs/issues/772 companion of https://github.com/o1-labs/snarkyjs/pull/887

### CODE REVIEW

File: crypto/elliptic_curve.ts [hunk 1]
```diff
@@ -30,6 +30,108 @@ const projectiveZero = { x: 1n, y: 1n, z: 0n };
...
+        const pow3 = F.power(x, 3n);
+        // a * x
+        const ax = F.mul(a, x);
+        // a^3 + ax + b
+        return F.add(F.add(pow3, ax), b);
+      }
+
+      const y = f(x);
+      const sqrtY = F.sqrt(y);
+      return F.isSquare(y) && sqrtY ? { x, y: sqrtY } : undefined;
```
[the remark this answers is not in the document]
Comment (mitschabaude) [reply crypto/elliptic_curve.ts:87]: on the other hand -- you test for `isSquare`, therefore the `sqrtY` has to be defined, so ignore the previous suggestion and use something like
```ts
const y = f(x);
if (!F.isSquare(y)) return undefined;
return { x, y: F.sqrt(y)! }
 ```
 (returning early to save computation)

File: crypto/elliptic_curve.ts [hunk 2]
```diff
@@ -30,6 +30,108 @@ const projectiveZero = { x: 1n, y: 1n, z: 0n };
...
+    }
+
+    function s_to_v_truncated(s: STuple): [bigint, bigint, bigint] {
```
Comment (mitschabaude) [inline crypto/elliptic_curve.ts:90]: can we use camel case everywhere for consistency? `sToVTruncated()`

**Review by mitschabaude — COMMENTED** [review]
Nice work @Trivo25! I have a couple of comments.

In addition to those, I think this should have a unit test which demonstrates that the TS implementation equals the OCaml one.

You can find similar thngs being done in other unit tests like `mina-signer/src/signature.unit-test.ts`.
Instead of hard-coded test inputs, it's best to use random field elements with our property testing framework, like here in the finite field unit test:
https://github.com/o1-labs/snarkyjs-bindings/blob/290d7b8fbcdc7ecba2f8981811ee958800bc4ad5/crypto/finite-field.unit-test.ts#L26-L27

File: crypto/poseidon.ts [hunk 3]
```diff
@@ -50,6 +57,23 @@ function createPoseidon(
     return state[0];
   }
 
+  function hashToGroup(input: bigint[]) {
+    let digest = hash(input);
+    let g = fieldToGroup(digest);
```
[the remark this answers is not in the document]
Comment (Trivo25) [reply crypto/poseidon.ts:62]: That's a very good point, I forgot about the curve-dependency

File: crypto/poseidon.ts [hunk 4]
```diff
@@ -50,6 +57,23 @@ function createPoseidon(
...
+    if (!g) return undefined;
```
[the remark this answers is not in the document]
Comment (Trivo25) [reply crypto/poseidon.ts:63]: puhh! I thought I forgot that 0's are falsy again (damn you JavaScript lol) - but it indeed does indeed read cleaner

**Review by mitschabaude — APPROVED** [review]
Nice! I'm approving in advance but have two more comments about the test

File: crypto/poseidon.unit-test.ts [hunk 5]
```diff
@@ -26,6 +30,24 @@ for (let i = 0; i < testVectors.length; i++) {
 
 console.log('poseidon implementation matches the test vectors! 🎉');
 
+for (let i = 0; i < 50; i++) {
```
[the remark this answers is not in the document]
Comment (Trivo25) [reply crypto/poseidon.unit-test.ts:33]: ahh nice :D I was looking for an option to do that:D

File: ocaml/lib/snarky_js_bindings_lib.ml [hunk 6]
```diff
@@ -1155,18 +1155,31 @@ type sponge =
   | Checked of Poseidon_sponge_checked.t
   | Unchecked of Poseidon_sponge.t
 
+let hash_array (xs : field_class Js.t Js.js_array Js.t) (is_checked : bool Js.t)
+    =
+  let input = Array.map (Js.to_array xs) ~f:of_js_field in
+  if Js.to_bool is_checked then Random_oracle.Checked.hash input
+  else Random_oracle.hash (Array.map ~f:to_unchecked input) |> Field.constant
```
Comment (rbonichon) [inline ocaml/lib/snarky_js_bindings_lib.ml:1162]:
```suggestion
  let a = Js.to_array xs in
  if Js.to_bool is_checked then Random_oracle.Checked.hash (Array.map a ~f:of_js_field)
  else Random_oracle.hash (Array.map a ~f:(fun x -> to_unchecked (of_js_field x)) |> Field.constant
```

One fewer iteration in the else case. Don't know if worth it ;-)

File: ocaml/lib/snarky_js_bindings_lib.ml [hunk 7]
```diff
@@ -1155,18 +1155,31 @@ type sponge =
...
-        if Js.to_bool is_checked then Random_oracle.Checked.hash input
+        if Js.to_bool is_checked then
+          Snark_params.Group_map.Checked.to_group input
         else
-          Random_oracle.hash (Array.map ~f:to_unchecked input) |> Field.constant
+          let x, y = Snark_params.Group_map.to_group (to_unchecked input) in
+          (Field.constant @@ x, Field.constant @@ y)
       in
-      to_js_field digest
+      to_js_group (fst digest) (snd digest)
```
Comment (rbonichon) [inline ocaml/lib/snarky_js_bindings_lib.ml:1182]:
```
if Js.to_bool is_checked then Snark_params.Group_map.Checked.to_group input
else
  let x, y = Snark_params.Group_map.to_group (to_unchecked input) in
  to_js_group (Field.constant x) (Field.constant y)
```
- the `@@` is unneeded
- don't need to construct `digest` to immediately destruct it through `fst` and `snd`

**Review by rbonichon — APPROVED** [review]
In general, LGTM.

Some improvements suggested in the OCaml code.

**Pull request merged**